### ISSUE TITLE
Increase the usage of compound assignment operators

### ISSUE THREAD

**Issue by elfring** [logisim-evolution/logisim-evolution#1328]
:eyes: Some source code analysis tools can help to find opportunities for improving software components.
:thought_balloon: I propose to [increase the usage of compound operators](https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.26 "Description for assignment operators") accordingly.

Would you like to integrate anything from a transformation result which can be generated by a command like the following?
(:point_right: Please check also for questionable change suggestions because of an evolving search pattern.)
```
lokal$ perl -p -i.orig -0777 -e 's#(?<target>\S+)\s*=\s*\k<target>[ \t]*(?<operator>[+\-*/%^]|&(?!&)|\|(?!\|)|<<|>>>?)#$+{target} $+{operator}=#gm' $(find ~/Projekte/logisim-evolution/lokal -name '*.java')
```

**Comment by davidhutchens** [comment]
I am closing this as there has been no interest shown. I also think this approach is more dangerous than it is worth.

**Comment by elfring** [comment]
>I also think this approach is more dangerous than it is worth.

How can [the application of compound operators](https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.26 "Description for assignment operators") be “dangerous” here? :thinking:

**Comment by davidhutchens** [comment]
It is dangerous as illustrated by the line you wrote just above your code. It needs to be checked by a person. And persons make mistakes.

Your process uses regular expressions and the code is a context free grammar. I didn't check your actual code but you can consider "a = a * b" replaced by "a *= b", but is "a = a * b + c" replaced by "a *= b + c"? There are lots of those kinds of things.

I've spent hours finding mistakes made by more clever automated code modification tools.

The benefit of making these changes is minimal. Some would even say we should go the other way and stop using compound assignments. I actually do use them, so I'm not in that camp. But it does suggest the benefit of your suggestion is small.

So the cost benefit analysis says to leave it alone.

**Comment by elfring** [comment]
>Your process uses regular expressions and the code is a context free grammar.

:eyes: The presented Perl script tries to demonstrate a few source code transformation possibilities.
It might be good enough for further clarification approaches.
:thought_balloon: Would anybody (besides me) dare to take another look at generated adjustments in a corresponding development repository?

**Comment by maehne** [comment]
I agree with @davidhutchens that such automated source code transformations need to be reviewed by humans and that nevertheless new errors can slip in. We made such experiences in the past. Nevertheless, I consider code clean ups that make the code more readable and maybe more efficient / generating less warnings worthwhile. However, to lower the risk of introducing errors, changes should be done with topical commits and keeping the scope/size of the PR at a reasonable size to facilitate the review. The longer a PR, the easier a problem slips through during review.
